node/src/components/small_network.rs

```diff
@@ -1259,116 +1263,185 @@ async fn setup_tls(
+/// or an error occurs.
 ///
 /// Initially sends a handshake including the `chainspec_hash` as a final handshake step.  If the
 /// recipient's `chainspec_hash` doesn't match, the connection will be closed.
-async fn message_sender<P>(
+#[allow(clippy::clippy::too_many_arguments)]
+async fn message_sender<P, REv>(
+    event_queue: EventQueueHandle<REv>,
     mut queue: UnboundedReceiver<Message<P>>,
     mut sink: SplitSink<FramedTransport<P>, Message<P>>,
+    mut shutdown: ShutdownReceiver,
     counter: IntGauge,
     handshake: Message<P>,
-) -> Result<()>
-where
+    peer_id: NodeId,
+    peer_address: SocketAddr,
+) where
     P: Serialize + Send + Payload,
+    REv: From<Event<P>>,
 {
-    sink.send(handshake).await.map_err(Error::MessageNotSent)?;
-    while let Some(payload) = queue.recv().await {
-        counter.dec();
-        // We simply error-out if the sink fails, it means that our connection broke.
-        sink.send(payload).await.map_err(Error::MessageNotSent)?;
-    }
+    // Start by sending the handshake.
+    if let Err(err) = sink.send(handshake).await.map_err(Error::MessageNotSent) {
+        event_queue
+            .schedule(
+                Event::OutgoingFailed {
+                    peer_id: Box::new(Some(peer_id)),
+                    peer_address: Box::new(peer_address),
+                    error: Box::new(Some(err)),
+                },
+                QueueKind::Network,
+            )
+            .await;
+        return;
+    };
+
+    loop {
+        tokio::select! {
+            // We are only cancelling this once, during shutdown, so we are not worried about
+            // cancellation safety here.
+            queue_recv_result = queue.recv() => {
+                match queue_recv_result {
+                    Some(payload) => {
+                        counter.dec();
+                        // We simply error-out if the sink fails, it means our connection broke.
+                        if let Err(err) = sink.send(payload).await.map_err(Error::MessageNotSent) {
+                            event_queue
+                            .schedule(
+                                Event::OutgoingFailed {
+                                    peer_id: Box::new(Some(peer_id)),
+                                    peer_address: Box::new(peer_address),
+                                    error: Box::new(Some(err)),
+                                },
+                                QueueKind::Network,
+                            )
+                            .await;
+                        return;
+                        }
+                    }
+                    None => {
+                        // The receive channel was closed. We're shutting down unclean?
+                        break;
+                    }
+                }
+            }
 
-    Ok(())
+            _ = shutdown.wait_for_shutdown() => {
```

Comment:
We're relying on tokio quite a bit when `futures` based stream-primitives would do the job.

Couldn't we do something like:
```
match queue.select(shutdown.wait_for_shutdown().into_stream()) {
    Either::Left((payload, _)) => {
...
    }, 
    Either::Right((value2, _)) => break,
...
}
```